### PR TITLE
Launch state machine, based on ros params

### PR DESCRIPTION
**Pull request by TimBuckers**
Launch all nodes via state machine
`roslaunch march_state_machine march_state_machine.launch`

The state machine determines which nodes should launch. This is based on the selected parameters in the rqt_launch_menu. Selected arguments such as enable ui are also handled by the state machine. These settings are passed on to the corresponding launch file.

Running in Simulation mode should work. However running on Exoskeleton is not fully function at this point.

close https://github.com/project-march/march-iv/pull/145

### CODE REVIEW

File: march_state_machine/launch/march_state_machine.launch [hunk 1]
```diff
@@ -1,6 +1,15 @@
 <launch>
     <node name="State_Machine" pkg="march_state_machine" type="StateMachine.py" output="screen"/>
     <node name="State_Machine_viewer" pkg="smach_viewer" type="smach_viewer.py" output="screen"/>
+    <!--<param name="march/launch_settings/run_configuration">true</param>-->
```
Comment (Ishadijcks) [inline march_state_machine/launch/march_state_machine.launch]: Remove commented code

File: march_state_machine/src/march_state_machine/launch_sm.py [hunk 2]
```diff
@@ -1,17 +1,71 @@
...
-
 from march_state_machine.states.EmptyState import EmptyState
+from march_state_machine.states.LaunchSettingsState import LaunchSettingsState
+from march_state_machine.states.LaunchState import LaunchState
 
 
 def create():
     sm_launch = smach.StateMachine(outcomes=['succeeded', 'failed'])
+    # Default launch parameters
+    sm_launch.userdata.launch_arguments = {'gazebo_ui': False}
```
Comment (Ishadijcks) [inline march_state_machine/src/march_state_machine/launch_sm.py]: Do we need to define default values here? Or should we do that in the launch menu ui?

Or does this need to happen for smach userdata?

File: march_state_machine/src/march_state_machine/launch_sm.py [hunk 3]
```diff
@@ -1,17 +1,71 @@
...
+                               transitions={'succeeded': 'HIGH VOLTAGE',
+                                            'failed': 'failed'})
+        smach.StateMachine.add('HIGH VOLTAGE', EmptyState(),
+                               transitions={'succeeded': 'IMOTIONCUBE',
+                                            'failed': 'failed'})
+        smach.StateMachine.add('IMOTIONCUBE', EmptyState(),
+                               transitions={'succeeded': 'ETHERCAT',
+                                            'failed': 'failed'})
+        smach.StateMachine.add('ETHERCAT', EmptyState(),
+                               transitions={'succeeded': 'RQT INPUT DEVICE',
```
Comment (Ishadijcks) [inline march_state_machine/src/march_state_machine/launch_sm.py]: When you have a sequence like this, the smach sequencecontainer is very nice to reduce boilerplate code.
http://wiki.ros.org/smach/Tutorials/Sequence%20container

File: march_rqt_launch_menu/src/march_rqt_launch_menu/launch_menu.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+from qt_gui.plugin import Plugin
+from python_qt_binding import loadUi
+from python_qt_binding.QtWidgets import QWidget
+from python_qt_binding.QtWidgets import QGroupBox
+from python_qt_binding.QtWidgets import QRadioButton
+from python_qt_binding.QtWidgets import QCheckBox
+from march_rqt_launch_menu.run_configuration_enum import run_configuration_enum
+
+
+class LaunchMenuPlugin(Plugin):
```
Comment (Ishadijcks) [inline march_rqt_launch_menu/src/march_rqt_launch_menu/launch_menu.py]: Amazing functionality ;)

File: march_state_machine/src/march_state_machine/states/LaunchSettingsState.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    """
+
+    def __init__(self):
+        smach.State.__init__(self, outcomes=['exoskeleton', 'simulation', 'failed'], io_keys=["launch_arguments"])
+
+    def execute(self, userdata):
+        self.start_rqt_launch_menu()
+
+        # Wait until the run configuration is chosen in the rqt_launch_menu
+        rate = rospy.Rate(2)
```
Comment (Ishadijcks) [inline march_state_machine/src/march_state_machine/states/LaunchSettingsState.py]: Why such a low rate?

File: march_state_machine/src/march_state_machine/states/LaunchSettingsState.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        return rospy.get_param("/march/launch_settings/run_configuration")
+
+    @staticmethod
+    def start_rqt_launch_menu():
+        """Create and execute the command to launch rqt_launch_menu"""
+        command = "roslaunch march_rqt_launch_menu march_rqt_launch_menu.launch"
+        popen = subprocess.Popen(command, shell=True)
+        state = popen.poll()
+        if state is None:
+            time.sleep(2)
```
Comment (Ishadijcks) [inline march_state_machine/src/march_state_machine/states/LaunchSettingsState.py]: Why?

File: march_state_machine/src/march_state_machine/states/LaunchSettingsState.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            return 'succeeded'
+        elif state < 0:
+            rospy.loginfo("Process terminated with error")
+            return 'failed'
+        elif state > 0:
```
Comment (Ishadijcks) [inline march_state_machine/src/march_state_machine/states/LaunchSettingsState.py]: What if state = 0? Or does that never happen?

File: march_state_machine/src/march_state_machine/states/LaunchState.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        if not self.should_execute_launch_file():
+            rospy.loginfo("Do not launch " + self.launch_file_name)
+            return 'succeeded'
+
+        rospy.loginfo("Launch " + self.package_name + " " + self.launch_file_name)
+        return self.execute_launch_file()
+
+    def should_execute_launch_file(self):
+        """Check if this launch file should be skipped or not depending on the ros params"""
+        if self.should_launch_key == "":
```
Comment (Ishadijcks) [inline march_state_machine/src/march_state_machine/states/LaunchState.py]: Should launch key is kind of weird name?

File: march_state_machine/src/march_state_machine/states/LaunchState.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        if not rospy.has_param("march/launch_settings" + self.should_launch_key):
+            # If the parameter is not set do not execute the launch file and give a warning.
+            rospy.logwarn("Expected ros parameter not found: march/launch_settings" + self.should_launch_key)
+            return False
+        # Simply return the value of the parameter as result.
+        return rospy.get_param("march/launch_settings" + self.should_launch_key)
+
+    def execute_launch_file(self):
+        """create a command that launches the launch file defined in the input of the method """
+        command = "roslaunch {0} {1}".format(self.package_name, self.launch_file_name)
```
Comment (Ishadijcks) [inline march_state_machine/src/march_state_machine/states/LaunchState.py]: How does this work?

File: march_state_machine/src/march_state_machine/states/LaunchState.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        for param_name, param_path in self.launch_arguments.iteritems():
+            if rospy.has_param("march/launch_settings" + param_path):
+                command += " {0}:={1}".format(param_name, rospy.get_param("march/launch_settings" + param_path))
+        rospy.loginfo(command)
+        popen = subprocess.Popen(command, shell=True)
+        state = popen.poll()
+        if state is None:
+            time.sleep(4)
```
Comment (Ishadijcks) [inline march_state_machine/src/march_state_machine/states/LaunchState.py]: Sleep?

File: march_state_machine/src/march_state_machine/states/LaunchSettingsState.py (hunk 7 above)
Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/states/LaunchSettingsState.py]: copied this code from a form, the documentation is a bit vague: https://docs.python.org/2/library/subprocess.html#subprocess.Popen.returncode
I'll change it to an else case.

File: march_state_machine/src/march_state_machine/states/LaunchState.py (hunk 9 above)
Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: what do you mean, {0} and {1} are the arguments defined in the format(...)

Comment (Ishadijcks) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: Ah never seen python string literals before.

This post seem to suggest that it's python 3 as it links to those docs.
https://stackoverflow.com/questions/6682806/what-does-0-mean-in-this-python-string/6682826

Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: I've python 2.7 and it works

File: march_state_machine/src/march_state_machine/states/LaunchState.py (hunk 8 above)
Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: `launch_is_optional_key` is this a better name?

File: march_state_machine/src/march_state_machine/launch_sm.py (hunk 2 above)
Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/launch_sm.py]: I removed them

File: march_state_machine/src/march_state_machine/states/LaunchSettingsState.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            rate.sleep()
+
+        return rospy.get_param("/march/launch_settings/run_configuration")
+
+    @staticmethod
+    def start_rqt_launch_menu():
+        """Create and execute the command to launch rqt_launch_menu"""
+        command = "roslaunch march_rqt_launch_menu " \
+                  "march_rqt_launch_menu.launch"
+        popen = subprocess.Popen(command, shell=True)
```
Comment (JitskedeVries) [inline march_state_machine/src/march_state_machine/states/LaunchSettingsState.py]: Popen is a weird name. What does it do?

Comment (JitskedeVries) [reply march_state_machine/src/march_state_machine/states/LaunchSettingsState.py]: Open the Process?

File: march_state_machine/src/march_state_machine/states/LaunchState.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        self.launch_file_name = launch_file_name
+        self.launch_arguments = launch_arguments
+        self.launch_is_optional_key = launch_is_optional_key
+        smach.State.__init__(self, outcomes=['succeeded', 'failed'],
+                             io_keys=["launch_arguments"])
+
+    def execute(self, userdata):
+        if not self.should_execute_launch_file():
+            rospy.loginfo("Do not launch " + self.launch_file_name)
+            return 'succeeded'
```
Comment (JitskedeVries) [inline march_state_machine/src/march_state_machine/states/LaunchState.py]: Why is succeeded returned when you should not launch a certain launch file?

File: march_state_machine/src/march_state_machine/states/LaunchState.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            "march/launch_settings" + self.launch_is_optional_key)
+
+    def execute_launch_file(self):
+        """create a command that launches the launch file
+        defined in the input of the method """
+        command = "roslaunch {0} {1}".format(self.package_name,
+                                             self.launch_file_name)
+        command = self.add_parameters_to_command(command)
+
+        popen = subprocess.Popen(command, shell=True)
```
Comment (JitskedeVries) [inline march_state_machine/src/march_state_machine/states/LaunchState.py]: Again Popen

File: march_state_machine/src/march_state_machine/states/LaunchState.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        for param_name, param_path in self.launch_arguments.iteritems():
+            if rospy.has_param("march/launch_settings" + param_path):
+                command += " {0}:={1}".format(param_name, rospy.get_param(
+                    "march/launch_settings" + param_path))
+        return command
+
+    def should_execute_launch_file(self):
+        """Check if this launch file should be skipped
+         or not depending on the ros params"""
+        if self.launch_is_optional_key == "":
```
Comment (JitskedeVries) [inline march_state_machine/src/march_state_machine/states/LaunchState.py]: This probably is why Gazebo had no key with it. But it is still optional right? When you launch it with only the exoskeleton, you do not want Gazebo to start necessarily.

File: march_state_machine/src/march_state_machine/launch_sm.py [hunk 15]
```diff
@@ -1,17 +1,96 @@
...
+            Sequence.add('WAIT ON SETTINGS', LaunchSettingsState())
+
+        smach.StateMachine.add('FIRST SEQUENCE', first_sequence,
+                               transitions={'exoskeleton': 'HARDWARE SEQUENCE',
+                                            'simulation': 'SIMULATION SEQUENCE'})
+
+        # This is the Exoskeleton path:
+        # The EmptyStates are placeholders for upcoming functional code
+        # that we use to discuss and demo the launch feature.
+        # @TODO(Tim) replace EmptyState placeholders
```
[the remark this answers is not in the document]
Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/launch_sm.py]: Yes TODO's in PR are normal, when the TODO is not needed for the feature that the PR addresses. Most of the time you also create a issue for the TODO. However, in this case it's too general.

File: march_state_machine/src/march_state_machine/launch_sm.py [hunk 16]
```diff
@@ -1,17 +1,96 @@
...
+        with simulation_sequence:
+            Sequence.add('FAKE SENSOR DATA',
+                         LaunchState("march_fake_sensor_data",
+                                     "march_fake_sensor_data.launch",
+                                     launch_is_optional_key="/simulation/fake_temperature_sensors"
+                                     ))
+            Sequence.add('GAZEBO',
+                         LaunchState("march_gazebo",
+                                     "march_world.launch",
+                                     launch_is_optional_key='',
```
[the remark this answers is not in the document]
Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/launch_sm.py]: Good point, I added comments to the constructor method of the LaunchState

File: march_state_machine/src/march_state_machine/states/LaunchSettingsState.py (hunk 11 above)
Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/states/LaunchSettingsState.py]: I added a comment

File: march_state_machine/src/march_state_machine/states/LaunchState.py (hunk 12 above)
Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: Good question, short answer:
Skipping the launch file is the desired behavior, therefore 'succeeded'

long answer:
The launch file is on purpose skipped, by skipping the launch file the state did function properly. When returning 'succeeded' the state machine will continue tot the next planned state. Returning 'aborted' indicates that something went wrong. This probably means the system will shutdown.

File: march_state_machine/src/march_state_machine/states/LaunchState.py (hunk 14 above)
Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: The absence of the launch_is_optional_key means this state is not optional. 
However if you look at the possible paths in the state machine you can see that:
- if you launch as 'simulation' the gazebo LaunchState will be launched (always)
- if you launch as 'exoskeleton' the gazebo LaunchState will not be launched (never)
So this means Gazebo is never optional for a specific launch state. 

I understand now that this is the most difficult part to understand. When the documentation framework is done this week. I will add general text explaining this behavior.

File: march_state_machine/src/march_state_machine/states/LaunchState.py (hunk 12 above)
Comment (Ishadijcks) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: To add to Tims explanation. The `'succeeded'` only indicates that this state has finished succesfully, which it has, even as it skipped the launchfile.

File: .travis.yml [hunk 17]
```diff
@@ -8,5 +8,5 @@ install:
 
 script:
   - nosetests .
-  - pycodestyle
+  - pycodestyle . --max-line-length 120
```
[the remark this answers is not in the document]
Comment (TimBuckers) [reply .travis.yml]: it now according to pep8 standaard (119 characters). I added the comment:
PEP8 is the style we are using, the upper limit is 119 characters

File: march_state_machine/src/march_state_machine/states/LaunchState.py (hunk 14 above)
Comment (Ishadijcks) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: @TimBuckers that can be done now :)

Comment (JitskedeVries) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: So this should be documented then.

Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: Working on it!

Comment (TimBuckers) [reply march_state_machine/src/march_state_machine/states/LaunchState.py]: done

**Pull request merged**